Client/Frontend/Browser/Playlist/Cache/PlaylistDownloadManager.swift

```diff
@@ -0,0 +1,485 @@
+        task.resume()
+        
+        delegate?.onDownloadStateChanged(id: item.pageSrc, state: .inProgress, displayName: delegate?.displayNames(for: asset.preferredMediaSelection) ?? "")
+    }
+    
+    func cancelDownload(item: PlaylistInfo) {
+        for (task, value) in activeDownloadTasks where item.pageSrc == value.id {
+            task.cancel()
+            activeDownloadTasks.removeValue(forKey: task)
+            break
+        }
+    }
+    
+    func downloadTask(for pageSrc: String) -> MediaDownloadTask? {
+        for (_, asset) in activeDownloadTasks where pageSrc == asset.id {
+            return asset
+        }
+
+        return nil
+    }
+    
+    // MARK: - URLSessionDownloadDelegate
+    
+    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
+        guard let task = task as? URLSessionDownloadTask,
+              let asset = activeDownloadTasks.removeValue(forKey: task) else { return }
+
+        if let error = error as NSError? {
+            switch (error.domain, error.code) {
+            case (NSURLErrorDomain, NSURLErrorCancelled):
+                guard let cacheLocation = delegate?.localAsset(for: asset.id)?.url else { return }
+
+                do {
+                    try FileManager.default.removeItem(at: cacheLocation)
+                    Playlist.shared.updateCache(pageSrc: asset.id, cachedData: nil)
+                } catch {
+                    log.error("Could not delete asset cache \(asset.name): \(error)")
+                }
+
+            case (NSURLErrorDomain, NSURLErrorUnknown):
+                fatalError("Downloading HLS streams is not supported on the simulator.")
+
+            default:
+                fatalError("Fatal Error: \(error.domain)")
+            }
+            
+            DispatchQueue.main.async {
+                self.delegate?.onDownloadStateChanged(id: asset.id, state: .invalid, displayName: "")
+            }
+        }
+    }
+    
+    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didWriteData bytesWritten: Int64, totalBytesWritten: Int64, totalBytesExpectedToWrite: Int64) {
+        
+        guard let asset = activeDownloadTasks[downloadTask] else { return }
+        let percentage = (Double(totalBytesWritten) / Double(totalBytesExpectedToWrite)) * 100.0
+        
+        DispatchQueue.main.async {
+            self.delegate?.onDownloadProgressUpdate(id: asset.id, percentComplete: percentage)
+        }
+    }
+    
+    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
+        guard let asset = activeDownloadTasks.removeValue(forKey: downloadTask) else { return }
+        
+        if let response = downloadTask.response as? HTTPURLResponse, response.statusCode == 302 || response.statusCode >= 200 && response.statusCode <= 299 {
+            
+            var fileExtension = ""
```

Comment:
Maybe we can just set default extension here as mp4 since we are checking if it is emoty and setting it in line 427

Reply:
Won't work because we want to detect based on Content-Type header and file extension first.
If that fails, it would be empty and we'd then check based on file header byte sequences.. Only if both of the above fails, we want to default to MP4.